runtime/lua/vim/shared.lua

```diff
@@ -1192,29 +1221,60 @@ function vim._with(context, f)
     if not vim.api.nvim_win_is_valid(context.win) then
       error('Invalid window id: ' .. context.win)
     end
+    -- TODO: Maybe allow it?
+    if context.buf and vim.api.nvim_win_get_buf(context.win) ~= context.buf then
+      error('Can not set both `buf` and `win` context.')
+    end
   end
 
-  -- Store original options
-  local previous_options ---@type table<string, any>
-  if context.options then
-    previous_options = {}
-    for k, v in pairs(context.options) do
-      previous_options[k] =
-        vim.api.nvim_get_option_value(k, { win = context.win, buf = context.buf })
-      vim.api.nvim_set_option_value(k, v, { win = context.win, buf = context.buf })
+  -- Decorate so that save-set-restore options is done in correct window-buffer
+  local callback = function()
+    -- Save and set context (NOTE: triggers `OptionSet` event)
+    local cache = { bo = {}, env = {}, go = {}, wo = {} } ---@type table<string, any>
+
+    -- Use specific order from possibly most to least intrusive
+    for _, scope in ipairs(scope_order) do
```

Comment:
I am not sure how to make it less confusing, as it is caching and setting inside for loops.

Here is an example of what it does if it would make it easier to follow. For `{ o = { winblend = 25 }, bo = { undolevels = 0 } }`, it does the following sequence:
- Process `o` context:
    - Get its "effective"/"cachable" context via `nvim_get_option_info2()`. The result is one of `bo`, `go`, or `wo`. It is `wo` for `winblend`.
    - Cache it via `cache.wo.winblend = cache.wo.winblend or vim.wo.winblend or vim.NIL`. The first `cache.wo.winblend` is there in order to write to cache only once (as the same option name can be supplied in three different `context` fields). The fallback to `vim.NIL` is there mostly for `env` context (as `vim.env` can return `nil`).
    - Cache its global value via `cache.go.winblend = cache.go.winblend or vim.go.winblend`. This is needed because later setting option might affect global values (like for `o` or `wo` contexts).
    - Set it with `vim.o.winblend = 25`.
- Process `bo` context:
    - `cache.bo.undolevels = cache.bo.undolevels or vim.bo.undolevels or vim.NIL`.
    - `cache.go.undolevels = cache.go.undolevels or vim.go.undolevels`.
    - `vim.bo.undolevels = 0`.
- Execute callback `f`.
- Restore cache:
    - `vim.bo.undolevels = cache.bo.undolevels`
    - `vim.wo.winblend = cache.wo.winblend`
    - `vim.go.undolevels = cache.go.undolevels` and `vim.go.winblend = cache.go.winblend`.

Reply:
Ah ok, that clears things up a bit. Hmm, yeah I can't come up with a simpler approach for the time being. Let's go with this 👍